pkg/yurttunnel/dns/dns.go

```diff
@@ -0,0 +1,463 @@
+	case NodeDelete:
+		return dnsctl.onNodeDelete(event.Obj.(*corev1.Node))
+	case ServiceAdd:
+		return dnsctl.onServiceAdd(event.Obj.(*corev1.Service))
+	case ServiceUpdate:
+		return dnsctl.onServiceUpdate(event.Obj.(*corev1.Service))
+	case ServiceDelete:
+		return dnsctl.onServiceDelete(event.Obj.(*corev1.Service))
+	case ConfigMapAdd:
+		return dnsctl.onConfigMapAdd(event.Obj.(*corev1.ConfigMap))
+	case ConfigMapUpdate:
+		return dnsctl.onConfigMapUpdate(event.Obj.(*corev1.ConfigMap))
+	case ConfigMapDelete:
+		return dnsctl.onConfigMapDelete(event.Obj.(*corev1.ConfigMap))
+	default:
+		return nil
+	}
+}
+
+func (dnsctl *coreDNSRecordController) handleErr(err error, event interface{}) {
+	if err == nil {
+		dnsctl.queue.Forget(event)
+		return
+	}
+
+	if dnsctl.queue.NumRequeues(event) < maxRetries {
+		klog.Infof("error syncing event %v: %v", event, err)
+		dnsctl.queue.AddRateLimited(event)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	klog.Infof("dropping event %q out of the queue: %v", event, err)
+	dnsctl.queue.Forget(event)
+}
+
+func (dnsctl *coreDNSRecordController) ensureCoreDNSRecordConfigMap() error {
+	_, err := dnsctl.kubeClient.CoreV1().ConfigMaps(constants.YurttunnelServerServiceNs).
+		Get(yurttunnelDNSRecordConfigMapName, metav1.GetOptions{})
+	if err != nil && apierrors.IsNotFound(err) {
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      yurttunnelDNSRecordConfigMapName,
+				Namespace: constants.YurttunnelServerServiceNs,
+			},
+			Data: map[string]string{
+				yurttunnelDNSRecordNodeDataKey: "",
+			},
+		}
+		_, err = dnsctl.kubeClient.CoreV1().ConfigMaps(constants.YurttunnelServerServiceNs).Create(cm)
+		if err != nil {
+			return fmt.Errorf("failed to create ConfigMap %v/%v, %v",
+				constants.YurttunnelServerServiceNs, yurttunnelDNSRecordConfigMapName, err)
+		}
+	}
+	return err
+}
+
+func (dnsctl *coreDNSRecordController) syncTunnelServerService() error {
+	klog.V(2).Info("sync tunnel server service as whole")
+	dnatPorts := util.GetConfiguredDnatPorts(dnsctl.kubeClient, strconv.Itoa(dnsctl.insecurePort))
+	return dnsctl.updateTunnelServerSvcDnatPorts(dnatPorts)
+}
+
+func (dnsctl *coreDNSRecordController) syncDNSRecordAsWhole() {
+	klog.V(2).Info("sync dns record as whole")
+
+	dnsctl.lock.Lock()
+	defer dnsctl.lock.Unlock()
+
+	ip, err := dnsctl.getTunnelServerIP(false)
+	if err != nil {
+		klog.Errorf("failed to sync dns record as whole, %v", err)
+		return
+	}
+
+	nodes, err := dnsctl.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("failed to sync dns record as whole, %v", err)
+		return
+	}
+
+	records := make([]string, 0, len(nodes))
+	for i := range nodes {
+		node := nodes[i]
+		if !isEdgeNode(node) {
+			ip, err = getNodeHostIP(node)
+			if err != nil {
+				klog.Errorf("failed to parse node address for %v, %v", node.Name, err)
+				continue
+			}
+		}
+		records = append(records, formatDNSRecord(ip, node.Name))
```

Comment:
records for edge nodes are frogotten?

Reply:
We will record whether it is an edge node or not, but the destination ip is different.